Add info routes to provide deployment configuration

rabbah opened the pull request:
Provide deployment configuration: routes to API paths, available runtime information, and action/trigger limits.

See #335 for details.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Controller.scala:
```diff
@@ -76,6 +89,17 @@ class Controller(
+        "support" -> "https://github.com/openwhisk/openwhisk/issues".toJson,
+        "build" -> config(WhiskConfig.whiskVersionDate).toJson,
+        "buildno" -> config(WhiskConfig.whiskVersionBuildno).toJson,
+        "api_paths" -> Array("/api/v1").toJson,
```
Any particular reason why you use `Array` over something simpler like `List`?

rabbah replied:
So you have something to pick at?
thank you, and good night.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/Controller.scala:
```diff
@@ -95,12 +120,14 @@ object Controller {
     private class ServiceBuilder(config: WhiskConfig, instance: Int, logging: Logging) extends Creator[Controller] {
-        def create = new Controller(config, instance, logging)
+        // this method is not reached unless ExecManifest was initialized successfully
+        val runtimes = ExecManifest.runtimesManifest.getOrElse(Runtimes(Set()))
```
Should we have it throw an Exception then (or handle the `None` case appropriately) instead of defaulting to an empty set?

rabbah replied:
there was an assert there which i lost in git surgery - you make a good point, might as well throw an exception in the getter for this.

csantanapr commented on ansible/roles/nginx/templates/nginx.conf.j2:
```diff
@@ -32,14 +32,18 @@ http {
+        }
+
+        # proxy to the web action path
+        location /web {
```
@rabbah Does having this `/web` changes the vanity URL? and do we need to update the docs?

Today the docs say:
>https://guest.openwhisk.host/public/index.html becomes a alias for https://openwhisk.host/api/v1/experimental/web/guest/public/index.html
and 
also proxy https://guest.openwhisk.host to https://openwhisk.host/api/v1/experimental/web/guest/public/index.html

Does this now becomes ?
https://guest.openwhisk.host/public/index.html  -> https://guest.openwhisk.host/web/public/index.html 
and
https://guest.openwhisk.host -> https://guest.openwhisk.host/web

rabbah replied:
I intend to back this change out.

markusthoemmes commented on common/scala/src/main/scala/whisk/core/entity/ExecManifest.scala:
```diff
@@ -87,7 +89,17 @@ protected[core] object ExecManifest {
+                "deprecated" -> deprecated.getOrElse(false).toJson,
+                "default" -> default.getOrElse(false).toJson,
+                "attached" -> attached.isDefined.toJson,
+                "requireMain" -> requireMain.getOrElse(false).toJson)
```
The inner `toJson` are not necessary, either construct the `JsObject` directly and drop the later `toJson.asJsObject` or drop all the inner `toJson`

rabbah commented:
@markusthoemmes @csantanapr this is ready for your comments now.

1. added swagger for the runtime manifest and deployment parameters
2. added tests to validate response against swagger

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala:
```diff
@@ -44,35 +44,34 @@ import RestApiCommons._
-            "buildno" -> buildno.toJson)
+    private def apiDocsUrl = {
+        if (apipath.endsWith("/")) s"$apipath$swaggerdocpath"
+        else s"$apipath/$swaggerdocpath"
```
Should we rather strongly type the `Path` vals to be `spray.http.Uri.Path` instead of `String`? Could prevent the "unsureness" here.

markusthoemmes commented on tests/src/test/scala/system/rest/ControllerInfoSchemaTests.scala:
```diff
@@ -0,0 +1,54 @@
+        val response = RestAssured.
+            given().
+            config(sslconfig).
+            get(getBaseURL("/"))
```
Note to self: Figure out how to use Akka-HTTP clients with modified ssl config to get rid of this thing.

rabbah replied:
good luck :)

rabbah commented:
I'm blocking this until #1903 and #2048 are merged.

rabbah commented on core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala:
```diff
@@ -71,7 +71,7 @@ class ContainerPool(
     // Eventually, we will have a more sophisticated warmup strategy that does multiple sizes
     private val defaultMemoryLimit = MemoryLimit(MemoryLimit.STD_MEMORY)
-    private val NODEJS6_IMAGE = Exec.imagename("nodejs:6")
+    private val NODEJS6_IMAGE = ExecManifest.runtimesManifest.manifests("nodejs:6").image
```
@markusthoemmes heads up

rabbah commented:
No longer blocked. Ready for reviews.
PG2/1367 🏃

rabbah commented:
PG2/1367 🔵

markusthoemmes commented on common/scala/src/main/scala/whisk/core/entity/ExecManifest.scala:
```diff
@@ -91,12 +105,93 @@ protected[core] object ExecManifest {
+    protected[entity] object ImageName {
+        protected val defaultImageTag = "latest"
+        private val componentRegex = """([a-z0-9._-]+)""".r
+        private val tagRegex = """([\w.-]{0,128})""".r
```
Do we really need to have those narrow rules in here? After all, we're copying docker's regexes are we? What do we gain by this?

rabbah replied:
Early short circuit. Request won't reach the invoker only to fail then.

markusthoemmes replied:
Ah, for user provided images?

rabbah replied:
Indeed the check is only applied for construction of black box execs.

rabbah replied:
I'll add a comment.

markusthoemmes commented on core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala:
```diff
@@ -195,6 +198,21 @@ protected[controller] class RestAPIVersion_v1()(
+    private val info = JsObject(
+        "description" -> "OpenWhisk API".toJson,
+        "min_cli_version" -> whiskConfig(whiskVersionDate).toJson,
+        "min_recommended_cli_version" -> whiskConfig(whiskVersionDate).toJson,
```
We need a new config parameter for this to make sense, no?

rabbah replied:
Yes but we have to version the cli first. I could take these out but since we keep getting bit by the cli out of date issues, this gives us a way to now have the cli check when it's out of date.

dubeejw replied:
Hmm, I like the idea. Maybe the CLI should be keeping track of the backend versions that it works with instead of the controller doing that job. Should the controller know that a CLI even exists?

dubeejw replied:
Also, not sure I like having two info routes since they both seem to share some of the same information (build, api, path, etc).

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/container/ContainerPool.scala:
```diff
@@ -659,7 +659,10 @@ class ContainerPool(
     private def getDockerImageName(action: WhiskAction)(implicit transid: TransactionId): String = {
         // only Exec instances that are subtypes of CodeExec reach the invoker
-        val imageName = ExecImageName.containerImageName(config.dockerRegistry, config.dockerImagePrefix, action.exec.asInstanceOf[CodeExec[_]], config.dockerImageTag)
+        val exec = action.exec.asInstanceOf[CodeExec[_]]
```
☹️

rabbah replied:
We need to introduce a new type in the invoker so that it is dealing with code exec can action vs the more generic exec. Maybe something for the new invoker?

dubeejw commented on core/controller/src/main/scala/whisk/core/controller/Controller.scala:
```diff
@@ -135,6 +150,18 @@ object Controller {
 
     def optionalProperties = EntitlementProvider.optionalProperties
 
+    private def info(config: WhiskConfig, runtimes: Runtimes) = JsObject(
```
Include link to public slack channel?

rabbah replied:
Want to suggest a property name for it?

dubeejw replied:
Maybe add two entries to the support field, one for git and one for slack. 

Ex: `"support" -> JsObject("slack" -> JsString(slackURL), "git" -> JsString(gitURL))`

dubeejw commented on core/controller/src/main/scala/whisk/core/controller/Controller.scala:
```diff
@@ -135,6 +150,18 @@ object Controller {
+        "support" -> "https://github.com/openwhisk/openwhisk/issues".toJson,
+        "build" -> config(WhiskConfig.whiskVersionDate).toJson,
+        "build_tag" -> config(WhiskConfig.whiskVersionBuildno).toJson,
+        "api_paths" -> List("/api/v1").toJson,
```
Thinking about api/v2. Maybe the version here should be passed in via a parameter.

rabbah commented:
@dubeejw thanks for the feedback.

dubeejw commented on core/controller/src/main/resources/infoswagger.json:
```diff
@@ -0,0 +1,137 @@
+        },
+        "support": {
+          "type": "object",
+          "description": "Where to file defects, feature requests, reach the community for support, etc."
```
Needs comma at end of line.

dubeejw commented on core/controller/src/main/scala/whisk/core/controller/Controller.scala:
```diff
@@ -135,6 +150,18 @@ object Controller {
 
+    private def info(config: WhiskConfig, runtimes: Runtimes, apis: List[String]) = JsObject(
+        "description" -> "OpenWhisk".toJson,
+        "support" -> JsObject(
```
Could also add a link to the StackOverflow items tagged with OpenWhisk (https://stackoverflow.com/questions/tagged/openwhisk).

rabbah replied:
Eh let's not go overboard. At least git and slack we control links we control.

dubeejw commented on core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala:
```diff
@@ -195,6 +195,19 @@ protected[controller] class RestAPIVersion_v1()(
+    /**
+     * Describes details of a particular API path.
+     */
+    private val info = JsObject(
```
Is there a test for getting info from api/v1?

rabbah replied:
nope dont see one.

csantanapr commented:
@rabbah this ready now?

rabbah commented:
It is if I take out the swagger test for the new route - it's an incorrect test.

csantanapr commented:
@rabbah As usual when you think this is completely reviewed and ready to be merge, just assigned to me and I can click the green merge button :-)

Split this PR into two or address the testing problem, all good to me.
